client/data/hosting/use-cache.ts

```diff
@@ -0,0 +1,140 @@
+import {
+	useQuery,
+	useMutation,
+	UseMutationOptions,
+	useQueryClient,
+	useIsMutating,
+} from '@tanstack/react-query';
+import { useCallback } from 'react';
+import wp from 'calypso/lib/wp';
+
+export const EDGE_CACHE_ENABLE_DISABLE_NOTICE_ID = 'edge-cache-enable-disable-notice';
+export const USE_EDGE_CACHE_QUERY_KEY = 'edge-cache-key';
+export const TOGGLE_EDGE_CACHE_MUTATION_KEY = 'set-edge-site-mutation-key';
+export const CLEAR_EDGE_CACHE_MUTATION_KEY = 'clear-edge-site-mutation-key';
+
+interface MutationVariables {
+	name: string;
+}
+
+interface MutationResponse {
+	message: string;
+}
+
+interface MutationError {
+	code: string;
+	message: string;
+}
+
+export const getEdgeCacheStatus = async ( siteId: number ) => {
+	const response = await wp.req.get( {
+		path: `/sites/${ siteId }/hosting/edge-cache/active`,
+		apiNamespace: 'wpcom/v2',
+	} );
+
+	return response;
+};
+
+export const setEdgeCache = async ( siteId: number, newStatus: boolean ) => {
+	const response = await wp.req.post( {
+		path: `/sites/${ siteId }/hosting/edge-cache/active`,
+		apiNamespace: 'wpcom/v2',
+		body: {
+			active: newStatus,
+		},
+	} );
+	return response;
+};
+
+export const purgeEdgeCache = async ( siteId: number ) => {
+	return await wp.req.post( {
+		path: `/sites/${ siteId }/hosting/edge-cache/purge`,
+		apiNamespace: 'wpcom/v2',
+	} );
+};
+
+export const useEdgeCacheQuery = ( siteId: number ) => {
+	return useQuery< boolean, unknown, boolean >( {
+		queryKey: [ USE_EDGE_CACHE_QUERY_KEY, siteId ],
+		queryFn: () => getEdgeCacheStatus( siteId ),
+		enabled: !! siteId,
+		select: ( data ) => {
+			return !! data;
+		},
+		meta: {
+			persist: false,
+		},
+	} );
+};
+
+interface MutationError {
+	code: string;
+	message: string;
+}
+
+export const useSetEdgeCacheMutation = (
```

Comment:
It's possible to use the hook in the command palette, but requires more refactoring.

For custom ReactQuery mutations, it's a matter of taste which parameters get passed as hook arguments and which get passed to the mutation function. The mutation hook has been accepting the site ID as a hook argument and the new cache state as a mutation function argument. That's pretty typical and I think in good taste. Usually you know the site ID up front when you instantiate the hook. You don't want to be passing it over and over again to the mutation function.

However the command palette code throws a spanner in the works. It's one of the few places in Calypso where the UI doesn't know what site it's referring to until later. And so in that case it makes sense to pass the both the side ID and cache state to the mutation function.

An updated `useSetEdgeCacheMutation` hook might look like this.

```ts
interface MutationError {
	code: string;
	message: string;
}

interface MutationVariables {
	siteId: number;
	active: boolean;
}

export const useSetEdgeCacheMutation = (
	options: UseMutationOptions< boolean, MutationError, MutationVariables > = {}
) => {
	const queryClient = useQueryClient();
	const mutation = useMutation< boolean, MutationError, MutationVariables >( {
		mutationFn: async ( { siteId, active } ) => {
			return setEdgeCache( siteId, active );
		},
		onMutate: async ( { siteId, active } ) => {
			await queryClient.cancelQueries( {
				queryKey: [ USE_EDGE_CACHE_QUERY_KEY, siteId ],
			} );
			const previousData = queryClient.getQueryData( [ USE_EDGE_CACHE_QUERY_KEY, siteId ] );
			queryClient.setQueryData( [ USE_EDGE_CACHE_QUERY_KEY, siteId ], active );
			return previousData;
		},
		onError( _err, { siteId }, prevValue ) {
			// Revert to previous settings on failure
			queryClient.setQueryData( [ USE_EDGE_CACHE_QUERY_KEY, siteId ], Boolean( prevValue ) );
		},
		onSettled: ( ...args ) => {
			const { siteId } = args[ 2 ];
			// Refetch settings regardless
			queryClient.invalidateQueries( {
				queryKey: [ USE_EDGE_CACHE_QUERY_KEY, siteId ],
			} );
			options?.onSettled?.( ...args );
		},
	} );

	const { mutate, ...rest } = mutation;

	const setEdgeCacheCallback = useCallback(
		( siteId: number, active: boolean ) => mutate( { siteId, active } ),
		[ mutate ]
	);

	return { setEdgeCache: setEdgeCacheCallback, ...rest };
};
```

I removed the `mutationKey` option to get it working, but that's because the special `isLoading` handling doesn't appear to be used anywhere and so I was able to remove that. Otherwise more shenanigans would be needed to track the same mutation across different pages.

Reply:
Thanks, @p-jackson ! This is exactly the feedback I was looking for 😊

I applied the refactor with 5c4eccd36d5